### PR TITLE
Resolve #179

### LINKED ISSUE #179: Fix Calendar shifting date
See #178

### PR DESCRIPTION
**Pull request by MikhailTymchukDX**
This is a proper fix for #178 when the Calendar changes a date in negative GMT environment.

### CODE REVIEW

**Comment by AlekseyMartynov** [comment]
* Could this be unit-tested?
* Any reason to change API instead of fixing it?
* It seems you add 2 global functions. Can you avoid this?

**Comment by MikhailTymchukDX** [comment]
1. Yes.
2. No. Reverted.
3. If we don't change API then we keep `Date.fromISO()` global, but make`isArray()` private.

**Comment by AlekseyMartynov** [comment]
> If we don't change API then we keep Date.fromISO() global, but makeisArray() private.

`Date.fromISO` is not global, it's of `Date.prototype`

File: AjaxControlToolkit.Jasmine/Suites/CalendarTests.aspx [hunk 1]
```diff
@@ -446,6 +446,20 @@
...
                 });
             });
+
+            describe("Parsing", function () {
+                beforeEach(function () {
+                    this.extender = $find(CALENDAR_EXTENDER_CLIENT_ID);
+                });
+
+                it("parses date correctly", function () {
+                    debugger;
```
Comment (AlekseyMartynov) [inline AjaxControlToolkit.Jasmine/Suites/CalendarTests.aspx]: :beetle:

File: AjaxControlToolkit/Scripts/Calendar.js [hunk 2]
```diff
@@ -1753,6 +1753,22 @@ Sys.Extended.UI.CalendarBehavior.prototype = {
         }
         // make sure we clean up the flag due to issues with alert/alt-tab/etc
         this._popupMouseDown = false;
+    },
+    parseDateSortableFormat: function (dateString) {
+        var result = new Date(60 * new Date(0).getTimezoneOffset() * 1e3),
+        chunks = dateString.replace("Z", "").split("T"),
```
Comment (AlekseyMartynov) [inline AjaxControlToolkit/Scripts/Calendar.js]: no need to replace "Z" for sortable format

File: AjaxControlToolkit/Scripts/Calendar.js [hunk 3]
```diff
@@ -1753,6 +1753,22 @@ Sys.Extended.UI.CalendarBehavior.prototype = {
...
+        date = /(\d{4})-(\d{2})-(\d{2})/.exec(chunks[0]),
+        time = /(\d{2}):(\d{2}):(\d{2})\.?(\d{0,7})?/.exec(chunks[1]);
+        result.setFullYear(Number(date[1]));
+        result.setMonth(Number(date[2]) - 1);
+        result.setDate(Number(date[3]));
+        if(time.length) {
+            result.setHours(Number(time[1]));
+            result.setMinutes(Number(time[2]));
+            result.setSeconds(Number(time[3]));
+            result.setMilliseconds(Number(String(time[4]).substr(0, 3)) || 0);
```
Comment (AlekseyMartynov) [inline AjaxControlToolkit/Scripts/Calendar.js]: "s" format doesn't contain millis, safe to skip

File: AjaxControlToolkit/Scripts/Calendar.js [hunk 4]
```diff
@@ -1753,6 +1753,22 @@ Sys.Extended.UI.CalendarBehavior.prototype = {
         }
         // make sure we clean up the flag due to issues with alert/alt-tab/etc
         this._popupMouseDown = false;
+    },
+    parseDateSortableFormat: function (dateString) {
+        var result = new Date(60 * new Date(0).getTimezoneOffset() * 1e3),
+        chunks = dateString.split("T"),
```
Comment (AlekseyMartynov) [inline AjaxControlToolkit/Scripts/Calendar.js]: please indent propertly the `var` block:
```
var a = 1,
    b= 2;
```

**Comment by MikhailTymchukDX** [comment]
@AlekseyMartynov Anything else?

**Comment by AlekseyMartynov** [comment]
Yes! :smile: 
Two more remarks;

1. `if(time.length)` is not necessary because the string always has time components
2. Instead of cryptic "`new Date(60 * new Date(0).getTimezoneOffset() * 1e3)`" I'd write:
```
new Date(
    Number(date[1]),
    Number(date[2]) - 1,
    Number(date[3]),
    Number(time[1]),
    Number(time[2]),
    Number(time[3])
);

```

**Comment by AlekseyMartynov** [comment]
P.S. And also it's possible to use a single Regex!

**Comment by MikhailTymchukDX** [comment]
@AlekseyMartynov Done.

**Comment by AlekseyMartynov** [comment]
:shipit: squash it!

**Pull request merged**